### ISSUE TITLE
`getCookie` and `setCookie` do not work on the server api

### ISSUE THREAD

**Issue by Mini-ghost** [nuxt/nuxt#24180]
### Environment

```
- Operating System: Darwin
- Node Version:     v20.9.0
- Nuxt Version:     3.8.1
- CLI Version:      3.9.1
- Nitro Version:    2.7.2
- Package Manager:  pnpm@8.10.2
- Builder:          -
- User Config:      modules, components, devtools, app, runtimeConfig, security
- Runtime Modules:  @unocss/nuxt@0.57.2, nuxt-security@1.0.0-rc.3
- Build Modules:    -
```

### Reproduction

[Reproduction](https://stackblitz.com/edit/nuxt-starter-wfc4xc?file=public/user.ts,app.vue,server/api/user.ts)

1. Open [Reproduction](https://stackblitz.com/edit/nuxt-starter-wfc4xc?file=public/user.ts,app.vue,server/api/user.ts)
2. Set the cookies `counter` to 15. 
3. Refresh page, and some result not correct;

### Describe the bug

**`server/api/example.ts`**

```ts
export default defineEventHandler((event) => {
  // Read counter cookie
  let counter = getCookie(event, 'counter');

  return {
    counter,
  };
});

```

**`app.vue`**

```vue
<script setup lang="ts">
const { data } = useFetch('/api/example');

const { data: data2 } = useAsyncData(() => {
  return $fetch('/api/example');
});

const nuxtApp = useNuxtApp();
const { data: data3 } = useAsyncData(() => {
  return nuxtApp.ssrContext.event.$fetch('/api/example');
});
</script>

<template>
  <div>
    <!-- Assume the `counter` on cookies is 15 -->
    data:
    <pre>{{ data }}</pre> 
    <!-- { "counter": "15" } -->

    data2:
    <pre>{{ data2 }}</pre>
    <!-- { } -->

    data3:
    <pre>{{ data3 }}</pre>
    <!-- { "counter": "15" } -->
  </div>
</template>
```

The results of currently using `useFetch` and `useAsyncData` are different. I would like to know if this is expected behavior. I did not notice this in the documentation.

### Additional context

This issue should be related to the implementation inside `useFetch`, as follows:
https://github.com/nuxt/nuxt/blob/2832f0114b3365c8e7aadbf05b444443db64f818/packages/nuxt/src/app/composables/fetch.ts#L146-L151

**Comment by danielroe** [comment]
The issue here is that `$fetch` will not pass cookies on by default if called on the server, to avoid leaking cookies. For this, we have `useRequestFetch()` which will return a 'bound' fetch instance that _does_ pass cookies on by default.

**Comment by Mini-ghost** [comment]
Thanks for your answer, I noticed some issues regarding `useRequestFetch` and I'm looking forward to them.

- #19815

But I'm not sure what the security issue is, is it a concern about sending cookies to external server? If this is the case, how about encapsulate `useFetch` internally into `globalThis.$fetch`?

**Comment by danielroe** [comment]
Yes, the security issue is that this would pass a cookie meant for _your API_ to any other third party API that you then call from that point.

In addition, `$fetch` is not bound to a given request and can be used anywhere. So there would be the danger of cookies from _some other user's request_ being passed. Instead, it's better to be explicit when we _do_ want to pass cookies on.

**Comment by Mini-ghost** [comment]
In my work, I've tried adding a `$request` method and combining the implementation from within `useFetch` into the code. It checks paths to decide whether to use a pre-configured fetch instance that handles passing cookies.

```ts
const $request: typeof globalThis.$fetch = (request, options) => {
  if (import.meta.client) {
    return globalThis.$fetch(request, options);
  }

  // Refer to `useFetch` internal implementation
  if (typeof request === 'string' && request.startsWith('/')) {
    const _$fetch = nuxtApp.ssrContext?.event?.$fetch ?? globalThis.$fetch;
    return _$fetch(request, options as any);
  }

  return globalThis.$fetch(request, options);
};

globalThis.$request = $request;
```

Is it possible to integrate this into the internal implementation of `$fetch`?"

**Comment by danielroe** [comment]
I would be careful, because that will share cookies between requests. `nuxtApp` is bound to a particular request, but this is defined as a global and therefore when you have multiple requests being handled at the same time, cookies will/might be shared between them.

**Comment by Mini-ghost** [comment]
This perhaps related to this feature #24813

**Comment by Mini-ghost** [comment]
Let's track in #24813
